features/mbedtls/targets/TARGET_CRYPTOCELL310/ccm_alt.c

```diff
@@ -0,0 +1,135 @@
+}
+
+void mbedtls_ccm_init( mbedtls_ccm_context *ctx )
+{
+    memset( ctx, 0, sizeof( mbedtls_ccm_context ) );
+}
+
+void mbedtls_ccm_free( mbedtls_ccm_context *ctx )
+{
+    mbedtls_zeroize( ctx, sizeof( mbedtls_ccm_context ) );
+}
+
+int mbedtls_ccm_setkey( mbedtls_ccm_context *ctx,
+                        mbedtls_cipher_id_t cipher,
+                        const unsigned char *key,
+                        unsigned int keybits )
+{
+    if( ctx == NULL )
+        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    if( cipher != MBEDTLS_CIPHER_ID_AES ||
+         keybits != 128 )
+    {
+        return ( MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE );
+    }
+
+    memcpy( ctx->cipher_key , key, keybits/8 );
+    ctx->keySize_ID = CRYS_AES_Key128BitSize;
+
+    return ( 0 );
+
+}
+
+/*
+ * Authenticated encryption or decryption
+ */
+
+int mbedtls_ccm_encrypt_and_tag( mbedtls_ccm_context *ctx, size_t length,
+                         const unsigned char *iv, size_t iv_len,
+                         const unsigned char *add, size_t add_len,
+                         const unsigned char *input, unsigned char *output,
+                         unsigned char *tag, size_t tag_len )
+
+{
+    CRYSError_t CrysRet = CRYS_OK;
+    /*
+     * Check length requirements: SP800-38C A.1
+     * Additional requirement: a < 2^16 - 2^8 to simplify the code.
+     * 'length' checked later (when writing it to the first block)
+     */
+    if( tag_len < 4 || tag_len > 16 || tag_len % 2 != 0 )
+        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    /* Also implies q is within bounds */
+    if( iv_len < 7 || iv_len > 13 )
+        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    if( length > 0xFFFFFFFF || add_len > 0xFFFFFFFF )
+        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    CrysRet =  CRYS_AESCCM( SASI_AES_ENCRYPT, ctx->cipher_key, ctx->keySize_ID,(uint8_t*)iv, iv_len,
+                            (uint8_t*)add, add_len,  (uint8_t*)input, length, output, tag_len, tag );
+    if( CrysRet != CRYS_OK )
+        return ( MBEDTLS_ERR_CCM_HW_ACCEL_FAILED );
+
+    return ( 0 );
+
+}
+
+/*
+ * Authenticated decryption
+ */
+int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+                      const unsigned char *iv, size_t iv_len,
+                      const unsigned char *add, size_t add_len,
+                      const unsigned char *input, unsigned char *output,
+                      const unsigned char *tag, size_t tag_len )
+
+{
+    CRYSError_t CrysRet = CRYS_OK;
+    /*
+     * Check length requirements: SP800-38C A.1
+     * Additional requirement: a < 2^16 - 2^8 to simplify the code.
+     * 'length' checked later (when writing it to the first block)
+     */
+    if( tag_len < 4 || tag_len > 16 || tag_len % 2 != 0 )
+        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    /* Also implies q is within bounds */
+    if( iv_len < 7 || iv_len > 13 )
+        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    if( length > 0xFFFFFFFF || add_len > 0xFFFFFFFF )
```

Comment:
I agree, but this was a review comment to verify that length are not bigger than 32 bits, on some platforms. I'll add the check

Reply:
If you are also targetting platform with more than 32 bits then that makes sense, but as I meant, this could raise warnings in 32 bit platforms because the comparison will never succeed

Reply:
Also, note that there are several of these comparisons in the PR. I pointed some of them out...